hack/cleanup/konnect_control_planes.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strings"
+	"time"
+
+	sdkkonnectgo "github.com/Kong/sdk-konnect-go"
+	sdkkonnectcomp "github.com/Kong/sdk-konnect-go/models/components"
+	sdkkonnectops "github.com/Kong/sdk-konnect-go/models/operations"
+	"github.com/go-logr/logr"
+	"github.com/samber/lo"
+
+	"github.com/kong/kong-operator/controller/konnect/ops"
+	"github.com/kong/kong-operator/test"
+)
+
+const (
+	konnectControlPlanesLimit     = int64(100)
+	timeUntilControlPlaneOrphaned = time.Hour
+
+	kindKonnectGatewayControlPlane = "KonnectGatewayControlPlane"
+)
+
+// cleanupKonnectControlPlanes deletes orphaned control planes created by the tests and their roles.
+func cleanupKonnectControlPlanes(ctx context.Context, log logr.Logger) error {
+	serverURL, err := canonicalizedServerURL()
+	if err != nil {
+		return fmt.Errorf("invalid server URL %s: %w", test.KonnectServerURL(), err)
+	}
+
+	// NOTE: The domain for global endpoints is overridden in cleanup.yaml workflow.
+	// See https://github.com/Kong/sdk-konnect-go/issues/20 for details
+	sdk := sdkkonnectgo.New(
+		sdkkonnectgo.WithSecurity(
+			sdkkonnectcomp.Security{
+				PersonalAccessToken: sdkkonnectgo.String(test.KonnectAccessToken()),
+			},
+		),
+		sdkkonnectgo.WithServerURL(serverURL),
+	)
+
+	orphanedCPs, err := findOrphanedControlPlanes(ctx, log, sdk.ControlPlanes)
+	if err != nil {
+		return fmt.Errorf("failed to find orphaned control planes: %w", err)
+	}
+	if err := deleteControlPlanes(ctx, log, sdk.ControlPlanes, orphanedCPs); err != nil {
+		return fmt.Errorf("failed to delete control planes: %w", err)
+	}
+
+	return nil
+}
+
+// findOrphanedControlPlanes finds control planes that were created by the tests and are older than timeUntilControlPlaneOrphaned.
+func findOrphanedControlPlanes(
+	ctx context.Context,
+	log logr.Logger,
+	c *sdkkonnectgo.ControlPlanes,
+) ([]string, error) {
+	response, err := c.ListControlPlanes(ctx, sdkkonnectops.ListControlPlanesRequest{
+		PageSize: lo.ToPtr(konnectControlPlanesLimit),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list control planes: %w", err)
+	}
+	if response.ListControlPlanesResponse == nil {
+		body, err := io.ReadAll(response.RawResponse.Body)
+		if err != nil {
+			body = []byte(err.Error())
+		}
+		return nil, fmt.Errorf("failed to list control planes, status: %d, body: %s", response.GetStatusCode(), body)
+	}
+
+	var orphanedControlPlanes []string
+	for _, ControlPlane := range response.ListControlPlanesResponse.Data {
+		if ControlPlane.Labels[ops.KubernetesKindLabelKey] != kindKonnectGatewayControlPlane {
```

Comment:
REVIEW: should we add more checks to check whether the CP is created in integration tests?

Reply:
Let's add those 👍 sdk-konnect-go uses labels like : https://github.com/Kong/sdk-konnect-go/blob/0deeb38e6d822b55ee2e0d4e705f484a7f314948/test/integration/funcs.go#L17-L23.

Let's add something similar to ensure that these particular CPs are from KO's tests.

Reply:
We already added label `konghq.com/test-id` in tests. I added the filter in the listing of cps.

Reply:
Added a label `operator-test-id` in integration tests.